File: 2024/day_12/solution.py
def calculate_price(ornaments: str) -> int:
  ornament_values = {
    "*": 1,
    "o": 5,
    "^": 10,
    "#": 50,
    "@": 100,
  }

  total_price = 0
  previous_value = 0

  for current in ornaments:
    # Check for invalid ornaments
    if current not in ornament_values:
      return None

    current_value = ornament_values[current]

    # Determine whether to add or subtract the value
    if previous_value < current_value:
      total_price += current_value - 2 * previous_value
    else:
      total_price += current_value

  previous_value = current_value

  return total_price

File: 2024/day_12/solution.ts
(() => {
  function calculatePrice(ornaments: string): number | undefined {
    const ornamentValues: { [key: string]: number } = {
      "*": 1,
      o: 5,
      "^": 10,
      "#": 50,
      "@": 100,
    };

    let totalPrice = 0;
    let previousValue = 0;

    for (const current of ornaments) {
      // Check for invalid ornaments
      if (!ornamentValues[current]) {
        return undefined;
      }

      const currentValue = ornamentValues[current];

      // Determine whether to add or subtract the value
      if (previousValue < currentValue) {
        totalPrice += currentValue - 2 * previousValue;
      } else {
        totalPrice += currentValue;
      }

      previousValue = currentValue;
    }

    return totalPrice;
  }

  console.log(calculatePrice("***")); // 3
  console.log(calculatePrice("*o")); // 4
  console.log(calculatePrice("o*")); // 6
  console.log(calculatePrice("*o*")); // 5
  console.log(calculatePrice("**o*")); // 6
  console.log(calculatePrice("o***")); // 8
  console.log(calculatePrice("*o@")); // 94
  console.log(calculatePrice("*#")); // 49
  console.log(calculatePrice("@@@")); // 300
  console.log(calculatePrice("#@")); // 50
  console.log(calculatePrice("#@Z")); // undefined
})();
